Pull request

hhelgason opened the pull request:
Added scripts that generate percentile netCDF files and a script that transfers these files to the NKN network. Also modified a few files in the "Run VIC" step to account for a 60 day run time.

Commit:
Adding shortwave radiation and specific humidity to the get_metdata step. Adding subdaily disaggregation step. Solves issue #3.

Commit:
Committing Changes to the meteorological step

The get_metdata.py script is simplified, along with other minor changes.

Commit:
Committing multiple changes to the Meteorological and Run VIC steps.

Committing changes to scripts that download met data, generate subdaily forcings and run VIC, along with the accompanying ecf and configuration template files.

Commit:
Minor edits to combine.py and create_global_file.py

Commit:
Minor edits to netcdf.ecf and regrid.py

- Extra line removed from netcdf.ecf
- Unused library import deleted from regrid.py

Commit:
Committing edits following code review:

-Removed FileType from argparse and edited args.config_name call
-Simpified the get_metdata.py script with a for loop for filling in thread URLs and downloading the met data
-Added a more descriptive variable name for the met_delay variable in get_metdata.py
-A few other minor edits

Commit:
Committing post processing scripts

- Added scripts that generate percentile netCDF files and a script that transfers these files to the NKN network.

Commit:
Committing post processing scripts

- Added scripts that generate percentile netCDF files and a script that transfers these files to the NKN network.

Commit:
Changing VIC run time from 1 day to 60 days

Committing changes to a few scripts to account for a 60 day VIC run each day.

Commit:
Changing VIC run time from 1 day to 60 days

Committing changes to a few scripts to account for a 60 day VIC run each day.

Commit:
Revert "Changing VIC run time from 1 day to 60 days"

This reverts commit 6fd44c143c072fc3eeb5211731889f29898ba423.

Commit:
Revert "Committing post processing scripts"

This reverts commit 0d3a4aa22b4a49bbf00483c3b9f5be11fe8108a6.

Commit:
Merge remote-tracking branch 'upstream/develop' into feature/helgason_develop

bartnijssen commented:
@hhelgason : Can you provide a description of what the PR is about?

hhelgason commented:
Yes, I'm going to re-base this one, I'll add some info in a minute.

bartnijssen commented:
Don't worry about the rebase. I think it'll be OK after we merge.

bartnijssen commented on tools/bin/ecflow/processes/main/meteorological/combine.py:
```diff
@@ -95,7 +94,7 @@ def main():
 
     # remove the tocel_combine scripts from run_dir
-    map(os.remove, lat_lons_list)
+    #map(os.remove, lat_lons_list)
```
Don't commit lines that are commented out. If the lines are no longer needed simply delete the lines

bartnijssen commented on tools/bin/ecflow/processes/main/meteorological/get_metdata.py:
```diff
@@ -24,26 +24,41 @@
 lastyear_date = datetime.now() - timedelta(days=365 + n_days)
-lastyear_num_day = lastyear_date.timetuple().tm_yday - 1
+lastyear_num_day = lastyear_date.timetuple().tm_yday - \
+    1  # python 0 start correction
```
Put the comment on a separate line before it and move the 1 back to the end of line 55 to improve readability

bartnijssen commented on tools/bin/ecflow/processes/main/meteorological/get_metdata.py:
```diff
@@ -203,35 +218,66 @@
+        # download metdata from http://thredds.northwestknowledge.net
+        met_dsets = dict()
+        for var in varnames:
+            url = ("http://thredds.northwestknowledge.net:8080" +
```
It would be good to read these paths from configuration files rather than hardcode them. I would leave as is for now, but open an issue to move all hardcoded paths to configuration files.

hhelgason replied:
Indeed - Issue opened.

bartnijssen commented on tools/bin/ecflow/processes/main/meteorological/get_metdata.py:
```diff
@@ -203,35 +218,66 @@
+            ds.attrs = globe_attrs
+
+            # place data in dict
+            met_dsets[var[0]] = ds
```
What is `var[0]`? Perhaps extend the comment a bit to explain what you are doing

bartnijssen commented on tools/bin/ecflow/processes/main/models/create_global_file.py:
```diff
@@ -22,16 +23,33 @@
-model_year = model_date[:4]
-model_month = model_date[5:7]
-model_day = model_date[8:10]
+start_year = vic_start_date[:4]
```
If it is a standard `YYYY-MM-DD` then you can simply use an existing date parser. For example:

``` python

from dateutil.parser import parse
datestr = '2000-03-04"
start = parse(datestr)
```
Then `start.year`, `start.month` and `start.day` have each of the date elements

bartnijssen commented on tools/bin/ecflow/processes/main/post_processing/sm_analysis.py:
```diff
@@ -0,0 +1,142 @@
+# Weibull Plotting Position
+q = np.zeros(num_pp)
+for i in range(1, num_pp + 1):
+    q[i - 1] = (i / (num_pp + 1.0))
```
You should do this as a single array operation:
```python
q = np.arange(1,num_pp+1)/(num_pp + 1)
```

bartnijssen reviewed (changes requested):
Add description of what is in the PR (only the most recent changes, don't worry about the earlier commits)

Commit:
Update get_metdata.py

- incorporating a date parser to parse model dates.

Commit:
Committing further edits following code review:

- Fixed formatting in get_metdata.py and improved documentation
- Simplified the Weibull plotting position calculations in sm_analysis.py, swe_analysis.py and tm_analysis.py
- Uncommented the line in combine.py that removes the tocel_combine scripts from run_dir

orianac commented on tools/bin/ecflow/processes/main/meteorological/tocel_combine.py:
```diff
@@ -46,13 +44,15 @@
         final_dir, '{DATA_LAT_LON}'), combined, delimiter='  ', fmt="%.5f")
 
-else:  # if we only downloaded yesterday's data we need to append it to the existing data
+else:  # if we only downloaded the last 60 days we need to append it to the existing data
```
Is this a safe way to do this? I wonder whether it could safer to have `elif full_year == None` (is that what you're expecting it to be?) and then break on `else`? Not necessary now, I'm just wondering whether you could start doing funny things wit the `existing_data` modifications if something ends up in `existing_data` that you aren't expecting to be there.

hhelgason replied:
The full_year variable is created in the get_metdata.py script, it's set to 'Year' if we need to download the whole year and 'Day' if we can download a shorter period (the 'Day' name comes from when we were downloading only one day, I need to find a new name since now we download either a full year or 60 days).  This is based on whether the directory for the full year data is empty or not. Here, if full_year is not 'Year', it must be 'Day' and thus we append to the existing data.

orianac commented on tools/bin/ecflow/processes/main/models/create_global_file.py:
```diff
@@ -22,16 +23,26 @@
+end = parse(vic_end_date)
+save_state = parse(vic_save_state)
+
+in_state = os.path.join(state_path, 'state.%s%s%s_00000.nc' %
```
Descriptive line for what you're accessing here?

orianac commented on tools/bin/ecflow/processes/main/post_processing/sm_analysis.py:
```diff
@@ -0,0 +1,140 @@
+# read in configuration file
+parser = argparse.ArgumentParser(description='Calculate SWE percentiles')
+parser.add_argument('config_file', metavar='config_file',
+                    help='configuration file')
```
More descriptive help than just a longer version of the variable name? Maybe point to a sample config file?

orianac commented on tools/bin/ecflow/processes/main/post_processing/sm_analysis.py:
```diff
@@ -0,0 +1,140 @@
+latitude = np.repeat(un_lat, num_lon)
+longitude = np.tile(un_lon, num_lat)
+
+# Weibull Plotting Position
```
Reference?

orianac commented on tools/bin/ecflow/processes/main/post_processing/sm_analysis.py:
```diff
@@ -0,0 +1,140 @@
+# Dictionary to DataFrame to Dataset
+df = pd.DataFrame(d, columns=["Latitude", "Longitude", "smpercentile"])
+a = df['smpercentile'].values
+new = a.reshape(195, 245)
```
I know the domain is fixed right now, but perhaps this could be not hardcoded and instead read from an existing netcdf file?

hhelgason replied:
Good catch! These values are actually read from the VIC output file in this script (lines 59 and 69, num_lat and num_lon).

orianac commented on tools/bin/ecflow/processes/main/post_processing/sm_analysis.py:
```diff
@@ -0,0 +1,140 @@
+a = df['smpercentile'].values
+new = a.reshape(195, 245)
+
+dsx = xr.Dataset({'smpercentile': (['lat', 'lon'], new)},
```
What is the significance of the `dsx` and `dsxx` variable names? Do the `x` and `xx` mean something I don't understand? If not, maybe a more intuitive variable name would be useful.

hhelgason replied:
The x in dsx stands for xarray - perhaps not very intuitive. I made this a bit more readable (by moving the 100* multiplication to where the Weibull plotting position is generated - thus getting rid of the latter variable name).

orianac commented on tools/bin/ecflow/processes/main/post_processing/swe_analysis.py:
```diff
@@ -62,9 +63,7 @@
-q = np.zeros(num_pp)
-for i in range(1, num_pp + 1):
-    q[i - 1] = (i / (num_pp + 1.0))
+q = np.arange(1,num_pp+1)/(num_pp + 1.0)
```
You use this elsewhere- maybe make a function out of it?

orianac commented on tools/bin/ecflow/processes/main/post_processing/tm_analysis.py:
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python
+"""
+creating tm percentiles
```
Clarify acronym in the header documentation?

orianac commented on tools/bin/ecflow/processes/main/post_processing/tm_analysis.py:
```diff
@@ -0,0 +1,139 @@
+        d.append(combine)
+
+# Dictionary to DataFrame to Dataset
+df = pd.DataFrame(d, columns=["Latitude", "Longitude", "tmpercentile"])
```
This routine appears in the other routines as well- maybe make a `dict_to_df_to_ds` function?

hhelgason replied:
Good idea. I need to make changes to this routine in my next pull request, will make sure to include a function for it.

bartnijssen commented:
@hhelgason : After you respond to @orianac comments I will go ahead and merge

Commit:
Committing further edits following code review:

- Minor fixes to create_global_file.py, sm_analysis.py, swe_analysis.py and tm_analysis.py

hhelgason commented:
@bartnijssen : Oriana's comments have been addressed
